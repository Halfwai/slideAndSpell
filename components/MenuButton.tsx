import React, { useRef, useEffect } from 'react';
import { Text, View, Dimensions, Animated, TouchableOpacity, StyleSheet } from 'react-native';

import { useFocusEffect } from 'expo-router';

import { RFPercentage } from "react-native-responsive-fontsize";

import MyAppText from '@/components/MyAppText';

import { COLOURS } from '@/constants/colours';

export default function MenuButton(props: { text: string, onPress: Function, delay: number, exitMenu: boolean }) {
    const buttonPosition = useRef(new Animated.Value(Dimensions.get('window').width)).current;
    const [colour, setColour] = React.useState("white");
    function slideIn() {
        Animated.timing(buttonPosition, {
            toValue: 0,
            duration: 500,
            useNativeDriver: true
        }).start();
    }

    function slideOut() {
        setTimeout(() => {
            Animated.timing(buttonPosition, {
                toValue: -Dimensions.get('window').width,
                duration: 500,
                useNativeDriver: true
            }).start(({ finished }) => {
                if (!finished) return;
                setColour("white");
            });
        }, props.delay);
    }

    useFocusEffect(() => {
        if (!props.exitMenu) {
            setTimeout(() => {
                slideIn();
            }, props.delay);
        }
    });


    useEffect(() => {
        if (props.exitMenu) {
            slideOut();
        }
    }, [props.exitMenu]);



    return (
        <View style={styles(colour).container}>
            <Animated.View
                style={[styles(colour).button, {
                    transform: [{ translateX: buttonPosition }],
                }]}
            >
                <TouchableOpacity
                    style={styles(colour).touchable}
                    onPress={() => {
                        props.onPress();
                        setColour(COLOURS.green);
                    }}
                >
                    <MyAppText style={styles(colour).text}>{props.text}</MyAppText>
                </TouchableOpacity>
            </Animated.View>
        </View>
    );
}


// Define the styles for the MenuButton component
const styles = (colour: string) => StyleSheet.create({
    container: {
        justifyContent: 'center',
        alignItems: 'center',
        marginVertical: 5,
    },
    touchable: {
        justifyContent: 'center',
        alignItems: 'center',
        backgroundColor: colour,
        borderColor: COLOURS.green,
        borderRadius: 10,
        borderWidth: 1,
        padding: 10,
    },
    button: {
        width: "80%",
        
    },
    text: {
        color: 'black',
        fontSize: RFPercentage(4),
        textAlign: 'center',
    }
});