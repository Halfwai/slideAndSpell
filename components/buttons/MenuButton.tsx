import React, { useRef, useEffect, useContext } from 'react';
import { View, Dimensions, Animated, TouchableOpacity, StyleSheet, ViewStyle } from 'react-native';
import { RFPercentage } from "react-native-responsive-fontsize";
import * as Haptics from 'expo-haptics';

// Import COLOURS
import { COLOURS } from '@/constants/colours';

// Import context
import { UserContext } from '@/utils/context';

// Import components
import MyAppText from '@/components/common/MyAppText';

// Setup props
interface MenuButtonProps {
    text: string,
    onPress: Function,
    delay: number,
    exitMenu: boolean,
    style?: { height: number | string | undefined },
    slideIn?: boolean,
}

export default function MenuButton(props: MenuButtonProps) {
    // Set button colour
    const [colour, setColour] = React.useState("white");
    const userContext = useContext(UserContext);

    // Set button position
    const buttonPosition = useRef(new Animated.Value(Dimensions.get('window').width)).current;

    // Function to slide the button in
    function slideIn() {
        Animated.timing(buttonPosition, {
            toValue: 0,
            duration: 500,
            useNativeDriver: true
        }).start();
    }

    // Function to slide the button out. Uses the delay props to trigger staggered slide out with other buttons
    function slideOut() {
        setTimeout(() => {
            Animated.timing(buttonPosition, {
                toValue: -Dimensions.get('window').width,
                duration: 500,
                useNativeDriver: true
            }).start(({ finished }) => {
                if (!finished) return;
                setColour("white");
            });
        }, props.delay);
    }

    // Slides the button out when one is pressed
    useEffect(() => {
        if (props.exitMenu) {
            slideOut();
        }
        if (!props.exitMenu) {
            setTimeout(() => {
                slideIn();
            }, props.delay);
        }
        return () => {
            buttonPosition.stopAnimation();
        }
    }, [props.exitMenu]);

    return (
        <View style={styles(colour).container}>
            <Animated.View
                style={[styles(colour).button, {
                    transform: [{ translateX: buttonPosition }],
                }]}
            >
                <TouchableOpacity
                    style={[styles(colour).touchable, props.style as ViewStyle]}
                    onPress={async () => {
                        // If vibrate is enabled, triggers haptics
                        if (userContext && userContext.vibrate) {
                            await Haptics.impactAsync(Haptics.ImpactFeedbackStyle.Light);
                        }
                        props.onPress();
                        setColour(COLOURS.green);
                    }}
                    testID="menuButton"
                >
                    <MyAppText style={styles(colour).text}>{props.text}</MyAppText>
                </TouchableOpacity>
            </Animated.View>
        </View>
    );
}


// Define the styles for the MenuButton component
const styles = (colour: string) => StyleSheet.create({
    container: {
        justifyContent: 'center',
        alignItems: 'center',
        marginVertical: 5,
        width: '100%',
    },
    touchable: {
        justifyContent: 'center',
        alignItems: 'center',
        backgroundColor: colour,
        borderColor: COLOURS.green,
        borderRadius: 10,
        borderWidth: 1,
        padding: 10,
    },
    button: {
        width: "80%",

    },
    text: {
        color: 'black',
        fontSize: RFPercentage(3.8),
        textAlign: 'center',
    }
});