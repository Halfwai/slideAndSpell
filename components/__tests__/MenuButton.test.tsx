import { fireEvent, render, act } from '@testing-library/react-native';
import MenuButton from '@/components/buttons/MenuButton';

beforeEach(() => {
    jest.useFakeTimers();
});

afterEach(() => {
    jest.clearAllMocks();
    jest.useRealTimers();
});


it(`MenuButton Component renders correctly`, async() => {
    const { toJSON } = render(
        <MenuButton 
            text="Test"
            onPress={() => {}}
            delay={1}
            exitMenu={false}
        />
    );
    expect(toJSON()).toMatchSnapshot();
});

it('MenuButton press fires correctly', async() => {
    const mockOnPress = jest.fn();
    const { getByTestId } = render(
        <MenuButton
            text="Test Button"
            onPress={mockOnPress}
            delay={100}
            exitMenu={false}
        />
    );
    const button = getByTestId('menuButton');
    fireEvent.press(button);
    expect(mockOnPress).toHaveBeenCalled();
});