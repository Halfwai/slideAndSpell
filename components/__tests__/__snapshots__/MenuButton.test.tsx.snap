// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MenuButton Component renders correctly 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "justifyContent": "center",
      "marginVertical": 5,
      "width": "100%",
    }
  }
>
  <View
    collapsable={false}
    style={
      {
        "transform": [
          {
            "translateX": 750,
          },
        ],
        "width": "80%",
      }
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "backgroundColor": "white",
          "borderColor": "#8BF0BC",
          "borderRadius": 10,
          "borderWidth": 1,
          "justifyContent": "center",
          "opacity": 1,
          "padding": 10,
        }
      }
      testID="menuButton"
    >
      <Text
        style={
          [
            {
              "fontFamily": "Quicksand",
            },
            {
              "color": "black",
              "fontSize": 51,
              "textAlign": "center",
            },
          ]
        }
      >
        Test
      </Text>
    </View>
  </View>
</View>
`;
